lib/runtime/src/pipeline/network/ingress/http_endpoint.rs

```diff
@@ -0,0 +1,549 @@
+        }
+    };
+    drop(handlers);
+
+    // Increment inflight counter
+    handler.inflight.fetch_add(1, Ordering::SeqCst);
+
+    // Extract tracing headers
+    let traceparent = TraceParent::from_axum_headers(&headers);
+
+    // Spawn async handler
+    let service_handler = handler.service_handler.clone();
+    let inflight = handler.inflight.clone();
+    let notify = handler.notify.clone();
+    let namespace = handler.namespace.clone();
+    let component_name = handler.component_name.clone();
+    let endpoint_name = handler.endpoint_name.clone();
+    let instance_id = handler.instance_id;
+
+    tokio::spawn(async move {
+        tracing::trace!(instance_id, "handling new HTTP request");
+        let result = service_handler
+            .handle_payload(body)
+            .instrument(tracing::info_span!(
+                "handle_payload",
+                component = component_name.as_ref(),
+                endpoint = endpoint_name.as_ref(),
+                namespace = namespace.as_ref(),
+                instance_id = instance_id,
+                trace_id = traceparent.trace_id,
+                parent_id = traceparent.parent_id,
+                x_request_id = traceparent.x_request_id,
+                x_dynamo_request_id = traceparent.x_dynamo_request_id,
+                tracestate = traceparent.tracestate
+            ))
+            .await;
+        match result {
+            Ok(_) => {
+                tracing::trace!(instance_id, "request handled successfully");
+            }
+            Err(e) => {
+                tracing::warn!("Failed to handle request: {}", e.to_string());
+            }
+        }
+
+        // Decrease inflight counter
+        inflight.fetch_sub(1, Ordering::SeqCst);
+        notify.notify_one();
+    });
+
+    // Return 202 Accepted immediately (like NATS ack)
+    (StatusCode::ACCEPTED, "")
+}
+
+/// Shared state for HTTP endpoint handler
+#[derive(Clone)]
+struct HttpEndpointState {
+    service_handler: Arc<dyn PushWorkHandler>,
+    instance_id: i64,
+    namespace: Arc<String>,
+    component_name: Arc<String>,
+    endpoint_name: Arc<String>,
+    system_health: Arc<Mutex<SystemHealth>>,
+    inflight: Arc<AtomicU64>,
+    notify: Arc<Notify>,
+}
+
+#[derive(Builder)]
+pub struct HttpEndpoint {
+    pub service_handler: Arc<dyn PushWorkHandler>,
+    pub cancellation_token: CancellationToken,
+    #[builder(default = "true")]
+    pub graceful_shutdown: bool,
+    #[builder(default = "DEFAULT_RPC_ROOT_PATH.to_string()")]
+    pub rpc_root_path: String,
+}
+
+impl HttpEndpoint {
+    pub fn builder() -> HttpEndpointBuilder {
+        HttpEndpointBuilder::default()
+    }
+
+    /// Get the RPC root path from environment or use default
+    fn get_rpc_root_path() -> String {
+        std::env::var("DYN_HTTP_RPC_ROOT_PATH")
+            .unwrap_or_else(|_| DEFAULT_RPC_ROOT_PATH.to_string())
+    }
+
+    pub async fn start(
```

Comment:
are we starting an htttp service per endpoint?

Reply:
resolved. i apparently could not resolve it when it or delete it when it was in the draft.  after deeper review, i can see that only one tcp or one http service is being created.  at least to the best of my understanding.